[Feb 4 merge] Review GitHub lessons 0-3 (#105)
pyOpenSci/lessons: github-git/3-fork-repo.md

---
jupytext:
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.16.4
kernelspec:
  display_name: Python 3 (ipykernel)
  language: python
  name: python3
---

(fork-repository)=

> # How to Fork a GitHub Repository 

Reviewer: This is a different workflow, and so might not be easy to teach along side the lesson you have now, but if all edits happen in the browser, navigating to the file to be edited *first* and clicking the edit pen button will take the user to an alternative `Fork this repository` page (only if you haven't yet forked).

Author: OH gosh - so true. I discovered this when creating the screen captures! 

What is so hard is that there are actually multiple potential paths here! I thought about an overview page (i have a [draft online here](https://www.pyopensci.org/lessons/github-git/ways-to-contribute.html#start-here-if-you-haven-t-yet-forked-the-repository) that also shows them the various paths in case they happen to do what you just mentioned! 

But we stick with "first principles," so to speak, in this lesson, which would be the traditional issue, fork, work, commit, pr route. 

I want to show them one way clearly. But then maybe have a resource page that shows alternative paths? 

what do you think about that?

Reviewer: Probably best to stick to one workflow. And what you have is the more general, probably seen as more "proper" as it is how you would make a fix locally.

Best to keep it simple. I'll just note that every PyOS sprint I have gone to, I've encouraged attendees to go straight for the edit button.

Author: Yup. ok cool. yes i think we can in the future show them those workflows as well (i've done that too in sprints). 
i've started this page below:

https://www.pyopensci.org/lessons/github-git/ways-to-contribute.html

to capture the various ways you can create a pr! That way we have that documented for sprints too (we can work on it together via a future review as well). i plan to open more pr's here with a fresh version of each lesson so we can review it as if it's new after i get this grant deliverable in today (its done!).

:::{admonition} What you will learn:
:class: tip

In this lesson, you will learn how to <kbd>fork</kbd> (or create a copy of) a GitHub repo into your own **GitHub.com** account. You can practice forking the pyOpenSci example repository. 
::: 

:::{admonition} Activity: Fork a repository and modify a file

**1. Fork the pyOpenSci practice GitHub repository**
*******

> Fork the [pyOpenSci example repo](https://github.com/pyOpenSci/pyos-demo-package-contribute). Remember that a fork is a copy of a repository that is owned by someone else or an organization that lives in your GitHub account.

Reviewer: ```suggestion
Fork the [pyOpenSci example repo](https://github.com/pyOpenSci/pyos-demo-package-contribute). Remember that a fork is a copy of a repository that is owned by someone else.
```

A mild suggestion, but I don't think a reader wanting to learn the basics of github is going to manage multiple namespaces under one GitHub account and trade PRs between them. This seem to distract from the (slightly simple) definition.

Reviewer 2: in training my lab on git (most are electrical engineers not software ppl), another point of confusion is "what's the difference between a fork and a branch," not sure if that's worth expanding on here bc not sure what level we're pitching this lesson at

> **2. Open the file that you proposed changes to in the issue you selected in the [how to identify an issue lesson](2-identify-issue).**

Reviewer: This point seems out of context of the rest of the page, and more like the starting point of edit-commit-files.

3. Click on the <kbd><i class="fa-solid fa-code-fork"></i> Fork</kbd> button and select your user account when it asks you where you want to fork the repo. 
4. Once you have forked the repo, you will have a copy in your account. Navigate to your repo page. The URL should look something like this:

`https://github.com/your-user-name/repo-name`

:::{figure} /images/github/fork-repo-animated.gif
:alt: alt 

To fork a repo, first, navigate to the repo you want to fork. Then click the **fork** button in the upper right-hand corner of your screen. You can then create a copy of this repo in your account.
:::

## Who owns a GitHub repository?

When a repository is stored on **GitHub.com**, it is assigned a unique URL (i.e. link on the **GitHub.com** website) that can be used to find the repository and access its files. While repositories on **GitHub.com** can be made either public or private, the default is public for free **GitHub** accounts.

> In either case (public or private), the URL links to a **GitHub** repository always follows the same format: 

Reviewer: ```suggestion
In either case (public or private), the URL links to a **GitHub** repository always follow the same format: 
```

`https://github.com/username/repository-name`

> The username is the username owner of the repo owner. The username can either be an individual such as `lwasser` (your **GitHub** username), or it can represent an organization such as `pyOpenSci`.

Reviewer: ```suggestion
The username is the owner of the repository. The username can either be an individual such as `lwasser` (your **GitHub** username), or it can represent an organization such as `pyOpenSci`.
```

> For example, the repository that you will work within this lesson is owned by `pyopensci.` Therefore, the URL looks like this:

Reviewer: ```suggestion
For example, the repository that you will work within this lesson is owned by `pyopensci`. Therefore, the URL looks like this:
```

`https://github.com/pyopensci/repo-name`

(fork-repo)=

> ## What is forking a GitHub repository?

Reviewer 2: similar comment re: structure as in section 2 of the lesson - two big ideas in this section: "what is a fork" and "how to make a fork"

currently

```
## How to fork a GitHub repository
## Who owns a GitHub repository?
## What is forking a GitHub repository?
## A few tips about forking a repo
```

could be

```
## What is a fork?
### ... info about what forks are
## How to fork a GitHub repository
### ... info on how to make a fork
```

Author: i will reorg some tomorrow. One thing I was considering here was put the action items at the top, considering people have limited attention spans these days. And then explanation below. 

I can't decide. We are living in a world of video shorts and such so i wondered what would happen if the "how to" was at the top and the "understand what you just did" came later. As an educator, it's not my preference, but do people online have the attention span to read it all, or do they want to skim and go back to learn?

thoughts?

Forking is useful because the fork remains linked to the original repository. This allows you to:
- Update your fork with changes from the original repository.
- Suggest changes to the original repository, which the owner can review and accept.

By forking, everyone collaborates on their own copies of the project, ensuring the original files stay intact. All changes are tracked in the file history and can be undone if needed. You can fork a repository directly from its main page on **GitHub.com**.

## A few tips about forking a repo

1. A GitHub (or GitLab) repository fork is a copy of the GitHub repository that you own in your account. The repository, however, is still connected to the parent repository on GitHub that you forked. This connection means that you can submit pull requests and changes to the repository using your fork.

It also means you can update your fork anytime, as the parent repository is updated to ensure it stays in sync.

> 2. If you have already forked the repository but some time has passed. You should consider updating or syncing your fork. GitHub has a sync button that you can use to do this (`pyopensci/repo-name`). This will ensure that all of the files in your repository are current.

Reviewer: ```suggestion
2. If you have already forked the repository but some time has passed. You should consider updating or syncing your fork. GitHub has a sync button that you can use to do this (<kbd>Sync fork</kbd>). This will ensure that all of the files in your repository are current.
```

Reviewer 2: omg first time seeing the `<kbd>` element

<kbd>Comtrol</kbd> + <kbd>Elt</kbd> + <kbd>Balete</kbd>

> 2. If you have already forked the repository but some time has passed. You should consider updating or syncing your fork. GitHub has a sync button that you can use to do this (`pyopensci/repo-name`). This will ensure that all of the files in your repository are current.

Reviewer 2: re: this, i think it would be worth having a "forks v. branches" subsection in "what is a fork?" section - along the lines of "a fork is an independent copy of a repository, you have a double of all the branches (or maybe just `main` depending on options), in your fork, so if they do work upstream, your version of `main` will get out of date." which would give this a place to live, as well as allow us to discuss stuff like why you should work on branches even in your fork so you can sync `main` when needed and leave links for further reading on rebasing/merging to keep a fork up to date.

Author: Yes definitely a good call. I started a graphic that captures this but will add a bit more text either today or next week. We can leave this open for longer if needed to really get the lessons right now that the grant deliverable is in. 
<img width="1277" alt="Screenshot 2025-01-31 at 12 59 36 PM" src="https://github.com/user-attachments/assets/e03f876d-72d0-41c9-8633-81a6dee791ee" />

the last deliverable is community review 🚀 and it makes everything better.

---
jupytext:
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.16.4
kernelspec:
  display_name: Python 3 (ipykernel)
  language: python
  name: python3
---

(fork-repository)=
# How to Fork a GitHub Repository 

:::{admonition} What you will learn:
:class: tip

In this lesson, you will learn how to <kbd>fork</kbd> (or create a copy of) a GitHub repo into your own **GitHub.com** account. You can practice forking the pyOpenSci example repository. 
::: 

:::{admonition} Activity: Fork a repository and modify a file

**1. Fork the pyOpenSci practice GitHub repository**
*******

> Fork the <a href="http://www.github.com/pyopensci/repo-here" target="_blank">pyOpenSci demo repository on GitHub</a>. Remember that a fork is a copy of a repository that is owned by someone else or an organization that lives in your GitHub account.

Author: ```suggestion
Fork the [pyOpenSci example repo](https://github.com/pyOpenSci/pyos-demo-package-contribute). Remember that a fork is a copy of a repository that is owned by someone else or an organization that lives in your GitHub account.
```

**2. Open the file that you proposed changes to in the issue you selected in the [how to identify an issue lesson](2-identify-issue).**
*******

> In the previous issue, you selected or identified something that you wanted to fix in a GitHub repo. Ideally, if this is your first contribution, that fix is small and could potentially be completed in the GitHub online interface. 
> 
> * Select the file that you decided to work on in GitHub
> * Click on the edit <kbd><i class="fa-solid fa-pencil"></i></kbd> button in the GitHub interface.
> * Make the edits that you proposed to make.
> * Hit the <kbd>commit</kbd> button to save your edits

Author: ```suggestion
```
The next lesson is about editing and commiting; this information doesn't belong here.
